core/src/main/java/org/springframework/security/access/hierarchicalroles/RoleHierarchyImpl.java

```diff
@@ -213,4 +222,65 @@ else if (roleName.equals(lowerRole.getAuthority())) {
 
 	}
 
+	/**
+	 * Builder for {@link RoleHierarchyImpl}. Note that the data structure used to build
+	 * the hierarchy is a {@link Map} so defining multiple roles using
+	 * {@link #role(String)} will only use the last entry.
+	 *
+	 * @author Federico Herrera
+	 * @since 6.3
+	 */
+	public static final class Builder {
+
+		private final Map<String, List<String>> roleBranches;
+
+		private Builder() {
+			this.roleBranches = new LinkedHashMap<>();
+		}
+
+		public RoleBranchBuilder role(String role) {
+			Assert.hasText(role, "role must not be empty");
+			return new RoleBranchBuilder(this, role);
+		}
+
+		/**
+		 * Creates a {@link RoleHierarchyImpl} and sets the built hierarchy using
+		 * {@link RoleHierarchyUtils#roleHierarchyFromMap(Map)}.
+		 * @return a {@link RoleHierarchyImpl}
+		 */
+		public RoleHierarchyImpl build() {
+			String roleHierarchyRepresentation = roleHierarchyFromMap(roleBranches);
+			RoleHierarchyImpl roleHierarchy = new RoleHierarchyImpl();
+			roleHierarchy.setHierarchy(roleHierarchyRepresentation);
+			return roleHierarchy;
+		}
+
+		public static final class RoleBranchBuilder {
+
+			private final Builder parentBuilder;
+
+			private final String role;
+
+			private RoleBranchBuilder(Builder parentBuilder, String role) {
+				this.parentBuilder = parentBuilder;
+				this.role = role;
+			}
+
+			/**
+			 * Implied role(s).
+			 * @param impliedRoles role name(s) implied by the role.
+			 * @return the same {@link Builder}
+			 * @throws IllegalArgumentException if <code>impliedRoles</code> is null or
+			 * empty or contains any null element.
+			 */
+			public Builder impliesAllOf(String... impliedRoles) {
```

Comment:
I think it would be better if this were called `implies`. The reason is that `allOf` is implied by the fact that it is a varargs.

I think the `allOf` prefix makes sense for things like `AuthorizationManagers` since you can have an `anyOf` there too, which makes the clarification necessary in that case.

Reply:
Looking back at my first review, I realize that one of my comments used `implies` and another used `impliesAllOf`. Sorry for potentially being confusing about my position.